qiskit/algorithms/observables_evaluator.py

```diff
@@ -0,0 +1,155 @@
+    except Exception as exc:
+        raise AlgorithmError("The primitive job failed!") from exc
+
+    variance_and_shots = _prep_variance_and_shots(estimator_job, len(expectation_values))
+
+    # Discard values below threshold
+    observables_means = expectation_values * (np.abs(expectation_values) > threshold)
+    # zip means and standard deviations into tuples
+    observables_results = list(zip(observables_means, variance_and_shots))
+
+    return _prepare_result(observables_results, observables)
+
+
+def _handle_zero_ops(
+    observables_list: list[BaseOperator | PauliSumOp | int],
+) -> list[BaseOperator | PauliSumOp]:
+    """Replaces all occurrence of operators equal to 0 in the list with an equivalent ``PauliSumOp``
+    operator."""
+    if observables_list:
+        zero_op = PauliSumOp.from_list([("I" * observables_list[0].num_qubits, 0)])
+        for ind, observable in enumerate(observables_list):
+            if observable == 0:
+                observables_list[ind] = zero_op
+    return observables_list
+
+
+def _prepare_result(
+    observables_results: list[tuple[complex, tuple[complex | None, int | None]]],
+    observables: ListOrDict[BaseOperator | PauliSumOp],
+) -> ListOrDict[tuple[complex, tuple[complex | None, int | None]]]:
+    """
+    Prepares a list of tuples of eigenvalues and (variance, shots) tuples from
+    ``observables_results`` and ``observables``.
+
+    Args:
+        observables_results: A list of tuples (mean, standard deviation).
+        observables: A list or a dictionary of operators whose expectation values are to be
+            calculated.
+
+    Returns:
+        A list or a dictionary of tuples (mean, (variance, shots)).
+    """
+
+    if isinstance(observables, list):
+        observables_eigenvalues = [None] * len(observables)
+        key_value_iterator = enumerate(observables_results)
+    else:
+        observables_eigenvalues = {}
+        key_value_iterator = zip(observables.keys(), observables_results)
+
+    for key, value in key_value_iterator:
+        if observables[key] is not None:
+            observables_eigenvalues[key] = value
+    return observables_eigenvalues
+
+
+def _prep_variance_and_shots(
+    estimator_result: EstimatorResult,
+    results_length: int,
+) -> list[tuple[complex | None, int | None]]:
+    """
+    Prepares a list of tuples with variances and shots from results provided by expectation values
+    calculations. If there is no variance or shots data available from a primitive, the values will
+    be set to ``0``.
+
+    Args:
+        estimator_result: An estimator result.
+        results_length: Number of expectation values calculated.
+
+    Returns:
+        A list of tuples of the form (variance, shots).
+    """
+    if not estimator_result.metadata:
+        return [(0, 0)] * results_length
+
+    results = []
+    for metadata in estimator_result.metadata:
+        if metadata and "variance" in metadata.keys() and "shots" in metadata.keys():
```

Comment:
For this logic can we split this so its not an `and`. I.e. if we have variance pass that back, If was have shots pass that back. (We could pass the whole metadata back too let users try and get any keys they want save extracting specifics here). E.g maybe extract and pass back variance so as caller does not have to manage its presence or not, as is done here, and pass back the whole metadata which would include shots and anything else